### DISCUSSION TITLE
TinTin++ 2.02.10

### DISCUSSION THREAD

**Discussion by scandum** [scandum/tintin#38] [Announcements]
```
split.c         Now allowing #showme {text} {-1} to place text on the input
                line without #split being enabled.

trigger.c       Changed #tick to lazily get the second argument.

files.c         Added the READ FILE [NAME] event, triggers after a file has
                been read. %0 holds the file name.

draw.c          Added color gradient support to #draw bar. To use it you
                must specify two 256 color foreground codes for the color
                field and based on the min;max value it will pick a color
                somewhere in between.

                #loop 1 9 i #draw <DDF> scroll bar 1 1 1 40 {$i;9;<faa><afa>}

draw.c          Added the BAR drawing type. Currently only horizontal bars
                are supported, but for forward compatibility you should draw
                bars using #draw horizontal bar {square} {<min>;<max>;[color]}
                The width of the bar will be that of the square coordinate.

trigger.c       Added support for 10 digit precision delays. The main
                utility of this is execution order.

                #delay 0.00003 #show A;#delay 0.00002 #show B

                Since delays are processed in 0.01 second intervals these will
                both trigger within 0.01 seconds, but the display order will
                be: B A rathern than A B.

path.c          Added MAPPING and RUNNING to #path get.

                MAPPING returns 1 when path mapping, otherwise 0.
                RUNNING returns the delay until the next run, otherwise 0.00.

                #path get <variable> without an argument will store the
                length, position, mapping, and running values in a table.

trigger.c       Storing the creation time of tickers and delays in arg4, can
                be obtained using #info tickers/delays save.

substitute.c    Added support to call @function{} as @{function}{}.

event.c         The SCREEN FOCUS event no longer requires #config mouse to
                be enabled.

math.c          % in #math now supports floating point modulo operations.

data.c          #info will save #info variables to $info[VARIABLES] rather
                than $info[VARIABLE]. Same for other lists.

list.c          Added #list {var} filter {keep} {remove}. Keep and remove
                must be a regex pattern, the keep regex is ignored if {} is
                used, the remove regex is optional. #list indexate must be
                used to filter tables.

utf8.c          Added #config charset cp949 for Korean.
```

**Comment by X1113** [comment]
When the server connection fails, the delay will also run in the next session

**Reply by X1113** [reply]
sorry ,is #ticker

**Reply by scandum** [reply]
Try putting the ticker in a file, and using `#session <name> <host> <port> <file>`.

The 4th argument of #session is only read on successful connect.

**Comment by bonginc** [comment]
Anyone else having color bleeds happen while connected? I'll be typing and the colors of the typed letters will change color.

**Reply by scandum** [reply]
Is this with #split enabled?

**Reply by bonginc** [reply]
Yes it is.

**Reply by scandum** [reply]
That's odd. One thing worth checking is whether it's a problem unique to the terminal emulator you are using.

If it happens with no script loaded and just #split enabled I could take a look at the mud that's causing this.

**Reply by bonginc** [reply]
I've tried another terminal and it does the same thing. The mud is legends of the jedi. legendsofthejedi.com 5656. I'm going to update and see if that fixes the issue.

**Reply by bonginc** [reply]
That didn't fix the issue but going back to 2.01.05 the color bleed doesn't happen. So I don't know what's going on.

**Reply by scandum** [reply]
Enabling `#config color_patch on` might fix the problem. I made a note on my todo list to look into a fix.

**Reply by bonginc** [reply]
That resolved the color issue. Thanks for your help.

**Comment by kitaelee** [comment]
Dear developer tintin++

Charset EUC-KR for Korean. please.

Thank you.

**Reply by scandum** [reply]
There's CP949 support, not sure if that's helpful?

I'll look into basic EUC-KR support, but it might be a while.

**Comment by kitaelee** [comment]
Thank you for your email.

I'm using CP949 to UTF8, but letters are broken.

I will wait for your EUC-KR support.

Thank you very much.

[image: image.png]

2021년 7월 23일 (금) 오후 7:41, scandum ***@***.***>님이 작성:

**Reply by eyehint** [reply]
charset cp949 is working well. 
but cp949toutf8 is not working. that is broken to special character.
if you want to use euc-kr,
you have to use screen. 
you run screen and ctrl + a and type ":encoding euc-kr"
and tt++ charset cp949, you can show Korean word.

**Comment by kitaelee** [comment]
Thank you very much.

😀😀😀😀😀

2021년 8월 26일 (목) 오후 5:17, Park Byeong-gwan ***@***.***>님이 작성:

**Comment by scandum** [comment]
Is cp949toutf8 completely broken, or is it only missing a few characters?

**Reply by eyehint** [reply]
Korean Working Well But special charecter borken 
![image](https://user-images.githubusercontent.com/6007348/131461241-50fde8c6-9380-449c-a9a4-4b673f6824e5.png)
![image](https://user-images.githubusercontent.com/6007348/131461441-83bbde7d-c39f-40a5-aeff-42f76be02dfb.png)

**Reply by scandum** [reply]
Thanks, that'll give me something to test. :)

**Comment by kitaelee** [comment]
only missing a few characters.

2021년 8월 26일 (목) 오후 9:21, scandum ***@***.***>님이 작성:

**Reply by scandum** [reply]
Ok, I'll make a note to take a look at it sometime.

**Comment by kitaelee** [comment]
Thank you very much. :)

2021년 8월 27일 (금) 오전 1:58, scandum ***@***.***>님이 작성:

**Comment by scandum** [comment]
Alright, I managed to find a more complete cp949 to unicode table, updated the beta with a fix.

https://mudhalla.net/tintin-beta.tar.gz

I only fixed cp949 to unicode conversion, if more complete unicode to cp949 conversion is needed I can take a look at that as well.

**Reply by eyehint** [reply]
okay, I'll test and feedback for you.
Thanks

**Reply by scandum** [reply]
I went ahead and updated input handling as well, so in the newest beta you should be able to send special characters to the mud, instead of just receiving them.

**Reply by eyehint** [reply]
Everything Work very well include Korean Chinese word.
But Halfwidth and fullwidth forms Issue exists
It can avoid in MAC Iterm2, to use ambiguous charecter setting.
but other client I have no idea
But I'm saticefied with tihs version
![image](https://user-images.githubusercontent.com/6007348/131833231-d8f745ff-bfb7-4324-bc56-8fbbcb1abb51.png)

**Reply by eyehint** [reply]
Many Terminal can enable ambiguous characers as wide option. 
It work well
![image](https://user-images.githubusercontent.com/6007348/131837605-ff6e6a5d-5210-4210-bb06-37e98fde3f04.png)

**Reply by scandum** [reply]
You might be able to fix that problem using:
```
#SUBSTITUTE {─} {──}
#SUBSTITUTE {━} {━━}
```

**Reply by eyehint** [reply]
It's perfect!!
Thank you!!
![image](https://user-images.githubusercontent.com/6007348/131851487-0e838b8c-36b0-4874-89e6-ff13421b5696.png)

**Reply by CryAngel1970** [reply]
to eyehint
Did you try it on Linux?
eyehint 님 사용하는 OS가 리눅스세요?
저는 윈도우에서 CygWin으로 tintin++을 컴파일 했는데 ctrl+a 도 안먹히고 그러네요.

**To the developer**
I'm using a tintin++ compiled from Window to Cygwin.
I still have a problem because of EUC-KR.
TINTIN++ now supports CP949TUTF8.
CP949 and EUC-KR are slightly different sets of characters.
It is not possible to convert directly from CP949 to EUC-KR.
UTF8 may be converted into EUC-KR, and EUC-KR may be converted into UTF8 again.

So I found source C that converts UTF-8 to EUC-KR on the website below.
[http://nlp.kookmin.ac.kr/data/unitoutf.html](url)

I hope EUC-KR is marked directly as EUC-KR, not UTF8 or CP949.

I tried the next method you told me.
you run screen and ctrl + a and type ":encoding euc-kr"
and tt++ charset cp949, you can show Korean word.

This method seems to be used on Linux.
I tried the above method in Windows 10, but it failed.

I hope to support EUC-KR for users who use TINTIN in Windows OS.

The Korean version of Windows is basically cp
The terminal is also cp949.
However, most Korean mud games are euc-kr.

EUC-KR cannot be converted to CP949 immediately.
UTF-8 may be converted to EUC-KR and CP949.

So, I think we can convert the EUC-KR code we received from the Korean mud game to UTF8 first, and then we can convert UTF8 to CP949.

I used Google Translator because I can't speak English. Translation may not be smooth.

**Reply by scandum** [reply]
@CryAngel1970 

Thanks for the information. I will see if I can offer a EUC-KR to UTF-8 translation.
 
Does setting `#config charset` to `BIG-5` or `GBK` allow for proper display if the terminal's character set is set to EUC-KR? If you use the mintty terminal (Used by WinTin++) with Cygwin it should allow for this.

![image](https://user-images.githubusercontent.com/52607428/210132477-b95094c8-ec0e-45e5-b60b-7e5f45f2e41e.png)

**Reply by CryAngel1970** [reply]
Thank you for your quick reply.
I installed CygWin and compiled TinTin.
I run Cygwin64 Terminal, and in minty Terminal state ./tt++.Executed exe.
I typed #config charset cp949.
I set Character set to eucKR in Terminal option, the letters are not broken.

But instead of working
[https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701?h=](https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701?h=)
try to use

I copied only the compiled tt++.exe to the newly created folder.
I created a batch file. (Wintin.bat)
Start wt -w 1 ./tt++ main.tin

Unfortunately, Windows Termianl does not have the option to set euc-kr.
The charset of Windows Terminal is cp949.
Of course, I can change it to utf8 by typing chcp 65001.

However, a problem arises because the mud is EUC-KR.
In euc-kr, 8000 words were added to cp949(MS949).

I don't want to install both Cygwin and tt++.
I hope euc-kr is supported when running only as an exe on Windows 10.
The reason is that it doesn't explain how to install Cygwin to someone who doesn't know tt++.
This is to share script files including the tt++.exe executable and main.tin.

---------------------------------------------------------------------------------------------------
EUC-KR = KSC5601(Korean) + KSC5636(English)

KSC5601 <-> UNICODE conversion code (using table)
[https://mungi.kr/101](https://mungi.kr/101)

PHP KSC5601 and UTF-8 class
[https://github.com/OOPS-ORG-PHP/KSC5601](https://github.com/OOPS-ORG-PHP/KSC5601)

**Reply by scandum** [reply]
I forgot to ask, do you have the address and port of a mud that supports euc-kr?

**Reply by CryAngel1970** [reply]
There are several muds, but the ones with the most users are the following ones.
jura.ga 9999
The jura.ga server may block access from abroad.
---------------------
mindjin.com 7450
mymud.iptime.org 4000
toox.co.kr 6025
52.79.89.136 4000
today.wo.tc 8050
oldmud.shop 9999
toox.co.kr 6000

Some of the servers above may also have blocked access from abroad.
------------------------------------------------------------------------------
Among the mud clients, there is one called yTin.
I know that yTin is also derived from tintin++.

yTin is known as a variant of tintin++ that can run on Windows.
I know that the yTin developer is Chinese, and that's why they support 2-byte characters.
[yTin github](https://github.com/MudRen/ytin)

There is no menu to set encoding in the yTin option menu.
However, the font is not broken.

![yTin option](https://i.imgur.com/8v3ITKN.jpg)

**Comment by kitaelee** [comment]
Thank you very very much.

I solved the problem.

now i enjoy mud games.

thank you very much.

2021년 7월 23일 (금) 오후 9:49, 맛돌이 ***@***.***>님이 작성:

**Comment by kitaelee** [comment]
Awesome, thanks!

2021년 9월 2일 (목) 오전 5:04, scandum ***@***.***>님이 작성:

**Comment by kitaelee** [comment]
네 리눅스사용중입니다

2022년 12월 31일 (토) 오전 8:59, Youn Wan-Shik ***@***.***>님이 작성:
